app/warden/magic_link_strategy.rb

```diff
@@ -1,15 +1,28 @@
 class MagicLinkStrategy < Warden::Strategies::Base
+  MAX_TOKEN_DURATION = 24.hours
+
   def valid?
     params[:token].present?
   end
 
   def authenticate!
     candidate = FindCandidateByToken.call(raw_token: params[:token])
 
-    if candidate
+    if candidate.present? && token_not_expired?(candidate)
+      expire_token(candidate)
       success!(candidate)
     else
       fail!
     end
   end
+
+private
+
+  def expire_token(candidate)
+    candidate.update!(magic_link_token: nil, magic_link_token_sent_at: nil)
+  end
+
+  def token_not_expired?(candidate)
+    Time.now < candidate.magic_link_token_sent_at + MAX_TOKEN_DURATION
```

Comment:
Avoids having to think about precedence:

 ```suggestion
     Time.now < (candidate.magic_link_token_sent_at + MAX_TOKEN_DURATION)
```